functional_tests/src/runtest/main.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2017 Google Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"log"
+	"os/exec"
+	"strings"
+	"syscall"
+)
+
+const FAILED = "FAILED"
+const PASSED = "PASSED"
+
+var verbose = flag.Bool("verbose", false, "Verbose logging")
+
+func main() {
+	testSpec := flag.String("test_spec", "", "Path to a yaml or json file containing the test spec")
+	flag.Parse()
+
+	if *testSpec == "" {
+		log.Fatal("--test_spec must be specified")
+	}
+
+	suite := LoadSuite(*testSpec)
+	doSetup(suite)
+	report := doRunTests(suite)
+	defer report()
+	doTeardown(suite)
+}
+
+func info(text string, arg ...interface{}) {
+	log.Printf(text, arg...)
+}
+
+func runCommand(name string, args ...string) (err error, stdout string, stderr string) {
+	if *verbose {
+		info("Running command: %v", append([]string{name}, args...))
+	}
+	cmd := exec.Command(name, args...)
+	var stdoutBuffer bytes.Buffer
+	var stderrBuffer bytes.Buffer
+	cmd.Stdout = &stdoutBuffer
+	cmd.Stderr = &stderrBuffer
+	err = cmd.Run()
+	stdout = stdoutBuffer.String()
+	stderr = stderrBuffer.String()
+	if *verbose {
+		commandOutput("STDOUT", stdout)
+		commandOutput("STDERR", stderr)
+	}
+	return
+}
+
+func commandOutput(name string, content string) {
+	if len(content) <= 0 {
+		return
+	}
+
+	info("%s>>%s<<%s", name, content, name)
+}
+
+func doSetup(suite Suite) {
+	info(">>> Setting up...")
+	for _, setup := range suite.Setup {
+		err, _, _ := runCommand(setup.Command[:1][0], setup.Command[1:]...)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func doTeardown(suite Suite) {
+	info(">>> Tearing down...")
+	for _, teardown := range suite.Teardown {
+		err, _, _ := runCommand(teardown.Command[:1][0], teardown.Command[1:]...)
+		if err != nil {
+			info(" > Warning: Teardown command failed: %s", err)
+		}
+	}
+}
+
+func doRunTests(suite Suite) func() {
+	info(">>> Testing...")
+	results := make(map[int]string)
+	passing := make(map[int]bool)
+	for index, test := range suite.Tests {
+		doOneTest(index, test, suite, results, passing)
+	}
+
+	report := func() {
+		info(">>> Summary:")
+		for index := range suite.Tests {
+			info(" > %s", results[index])
+		}
+
+		for index := range suite.Tests {
+			if !passing[index] {
+				log.Fatalf("Some test has failed.")
```

Comment:
Could you also keep track of the specific test and its error (or does that exist within the `suite` construct)? Would make the reporting a little bit more useful.